### PR TITLE
feat:Deadlock caused by unlocking a non-existing lock

### PR DESCRIPTION
**Pull request by yudoutingle**
## Description
feat:Deadlock caused by unlocking a non-existing lock。
for example， minio-6 locks minio-5 at startup, but minio-5 restarts before minio-6 unlocked,
causing minio-6's lock to be lost, which leads to minio-6 deadlock.
In fixed code, it will return true if lock not exits.

issues :  https://github.com/minio/minio/issues/15519

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Optimization (provides speedup with no functional changes)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [ ] Documentation updated
- [ ] Unit tests added/updated

### CODE REVIEW

File: cmd/local-locker.go [hunk 1]
```diff
@@ -127,16 +127,17 @@ func (l *localLocker) Unlock(_ context.Context, args dsync.LockArgs) (reply bool
...
+		if lri, ok := l.lockMap[resource]; ok {
+			if !l.canTakeUnlock(resource) {
+				// Unless it is a write lock reject it.
+				err = fmt.Errorf("unlock attempted on a read locked entity: %s", resource)
+				continue
+			}
 			reply = l.removeEntry(resource, args, &lri) || reply
+		} else {
+			// return true if lock not exits.
+			reply = true
```
Comment (harshavardhana) [inline cmd/local-locker.go]: This is wrong @yudoutingle you should never return true for Unlocks() on non-existent resources. This shows bugs here is the simplest fix

```diff
diff --git a/cmd/local-locker.go b/cmd/local-locker.go
index f89efcc7c..8eff129df 100644
--- a/cmd/local-locker.go
+++ b/cmd/local-locker.go
@@ -129,12 +129,12 @@ func (l *localLocker) Unlock(_ context.Context, args dsync.LockArgs) (reply bool
        err = nil
 
        for _, resource := range args.Resources {
-               if !l.canTakeUnlock(resource) {
+               lri, ok := l.lockMap[resource]
+               if ok && !isWriteLock(lri) {
                        // Unless it is a write lock reject it.
                        err = fmt.Errorf("unlock attempted on a read locked entity: %s", resource)
                        continue
                }
-               lri, ok := l.lockMap[resource]
                if ok {
                        reply = l.removeEntry(resource, args, &lri) || reply
                }
```

**Review by harshavardhana — CHANGES_REQUESTED** [review]

**Comment by harshavardhana** [comment]
Also this deadlock will not happen in recent releases due to the "Refresh()" routine 

> for example， minio-6 locks minio-5 at startup, but minio-5 restarts before minio-6 unlocked,

Your assessment here is based on an older release that had the problem of deadlocks minio-6 `unlock` now automatically refreshes the stale state. 

However, your fix is okay to take as an ice change.

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :heavy_check_mark: |
| mint-pools.sh | :heavy_check_mark: |
Deleting image on docker hub
Deleting image locally

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :heavy_check_mark: |
| mint-pools.sh | :heavy_check_mark: |
Deleting image on docker hub
Deleting image locally

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :heavy_check_mark: |
| mint-pools.sh | :heavy_check_mark: |
Deleting image on docker hub
Deleting image locally

**Review by klauspost — APPROVED** [review]
LGTM

**Review by harshavardhana — APPROVED** [review]

**Pull request merged**